widget/completion_entry.go

```diff
@@ -0,0 +1,197 @@
+type CompletionEntry struct {
+	widget.Entry
+	popupMenu     *widget.PopUp
+	navigableList *navigableList
+	Options       []string
+	pause         bool
+	itemHeight    float32
+}
+
+// NewCompletionEntry creates a new CompletionEntry which creates a popup menu that responds to keystrokes to navigate through the items without losing the editing ability of the text input.
+func NewCompletionEntry(options []string) *CompletionEntry {
+	c := &CompletionEntry{Options: options}
+	c.ExtendBaseWidget(c)
+	return c
+}
+
+// ShowCompletion displays the completion menu
+func (c *CompletionEntry) ShowCompletion() {
+	if c.pause {
+		return
+	}
+
+	if c.navigableList == nil {
+		c.navigableList = newNavigableList(c.Options, &c.Entry, c.setTextFromMenu, c.HideCompletion)
+	}
+	holder := fyne.CurrentApp().Driver().CanvasForObject(c)
+
+	if c.popupMenu == nil {
+		c.popupMenu = widget.NewPopUp(c.navigableList, holder)
+	}
+	c.popupMenu.Resize(c.maxSize())
+	c.popupMenu.ShowAtPosition(c.popUpPos())
+	holder.Focus(c.navigableList)
+}
+
+// HideCompletion hides the completion menu.
+func (c *CompletionEntry) HideCompletion() {
+	if c.popupMenu != nil {
+		c.popupMenu.Hide()
+	}
+}
+
+// SetOptions set the completion list with itemList and update the view.
+func (c *CompletionEntry) SetOptions(itemList []string) {
+	c.Options = itemList
+	c.Refresh()
+}
+
+// Refresh the list to update the options to display.
+func (c *CompletionEntry) Refresh() {
+	c.Entry.Refresh()
+	if c.navigableList != nil {
+		c.navigableList.SetOptions(c.Options)
+	}
+}
+
+// calculate where the popup should appear
+func (c *CompletionEntry) popUpPos() fyne.Position {
+	entryPos := fyne.CurrentApp().Driver().AbsolutePositionForObject(c)
+	return entryPos.Add(fyne.NewPos(0, c.Size().Height))
+}
+
+// calculate the max size to make the popup to cover everything below the entry
+func (c *CompletionEntry) maxSize() fyne.Size {
+	cnv := fyne.CurrentApp().Driver().CanvasForObject(c)
+
+	if c.itemHeight == 0 {
+		// set item height to cache
+		c.itemHeight = c.navigableList.CreateItem().MinSize().Height
+	}
+
+	listheight := float32(len(c.Options))*(c.itemHeight+2*theme.Padding()+theme.SeparatorThicknessSize()) + 2*theme.Padding()
+	canvasSize := cnv.Size()
+	entrySize := c.Size()
+	if canvasSize.Height > listheight {
+		return fyne.NewSize(entrySize.Width, listheight)
+	}
+
+	return fyne.NewSize(
+		entrySize.Width,
+		canvasSize.Height-c.Position().Y-entrySize.Height-theme.InputBorderSize()-theme.Padding())
+}
+
+// Move changes the relative position of the select entry.
+//
+// Implements: fyne.Widget
+func (c *CompletionEntry) Move(pos fyne.Position) {
+	c.Entry.Move(pos)
+	if c.popupMenu != nil {
+		c.popupMenu.Resize(c.maxSize())
+		c.popupMenu.Move(c.popUpPos())
+	}
+}
+
+// Prevent the menu to open when the user validate value from the menu.
+func (c *CompletionEntry) setTextFromMenu(s string) {
+	c.pause = true
+	c.Entry.SetText(s)
```

Comment:
Maybe the cursor should be shown at the end of the text when the user selects an option?:
```go
...
c.Entry.SetText(s)
rs := []rune(s)
c.Entry.CursorColumn = len(rs)
...
```

Reply:
I'm not sure about this. I did it before but is it the wanted behavior and is it common ? If yes, it's a simple line of code to add 😄

Reply:
Not sure either. Any thoughts? @Jacalz @andydotxyz

Reply:
I think that is up to this widget really... I would expect that makes sense, but I don't think there is a precedent to look at inside the toolkit. Other than `test.Type` I suppose, where we simulate a user typing.

Reply:
OK, I made the changes - this will be here on the next push, thanks 😉

Reply:
Done https://github.com/fyne-io/fyne-x/pull/11/commits/7504cfc4ffba90b954f425589b6ff5604bded166

Reply:
@metal3d You should use the length of the runes instead of bytes, as I mentioned above. Because Entry works internally with runes not with bytes

Reply:
My bad, I don't know why I did it... Sorry

Reply:
Fixed: https://github.com/fyne-io/fyne-x/pull/11/commits/839d80dedc3e98730d202b8d151b80be71ceb303

Reply:
> My bad, I don't know why I did it... Sorry

No problems :)